app/src/main/java/com/android/streetworkapp/ui/park/ParkOverview.kt

```diff
@@ -253,6 +282,59 @@ fun RatingDialog(showDialog: MutableState<Boolean>) {
   }
 }
 
+/**
+ * Verifies that the current state is correct and then rates the park.
+ *
+ * @param context The context of the application.
+ * @param park The park to rate.
+ * @param user The user who is rating the park.
+ * @param starRating The rating value.
+ * @param parkViewModel The park view model.
+ */
+fun handleRating(
+    context: Context?,
+    park: Park?,
+    user: User?,
+    starRating: Int,
+    parkViewModel: ParkViewModel?
+) {
+  Log.d("ParkOverview", "handleRating: {park=$park ; user=$user ; rating=$starRating")
+
+  if (user == null) {
+    if (context != null)
+        Toast.makeText(context, "User not found, could not rate park", Toast.LENGTH_SHORT).show()
+  } else if (park == null) {
+    if (context != null)
+        Toast.makeText(context, "Park not found, could not rate park", Toast.LENGTH_SHORT).show()
+  } else if (starRating < 1 || starRating > 5) {
+    if (context != null)
+        Toast.makeText(context, "Invalid rating value, could not rate park", Toast.LENGTH_SHORT)
+            .show()
+  } else if (parkViewModel == null) {
+    if (context != null)
+        Toast.makeText(context, "Error with view model, could not rate park", Toast.LENGTH_SHORT)
+            .show()
+  } else {
+    Log.d("ParkOverview", "handleRating: Adding rating to park")
+    if (context != null) Toast.makeText(context, "Rating submitted", Toast.LENGTH_SHORT).show()
```

Comment:
Instead of calling context != null everywhere call it once at the start

Reply:
### Why the null checks ?
The reason for the null checking is that each `Toast()` instance require a `context`. Because I want to test this component (`handleRating`) and it is not a **@composable** function, it has no associated `context` by itself (cannot fetch it using `localContext.current` for instance). Meaning I need to **pass it from the UI which calls it.** 

At the same time when I test the component, I cannot give it a non-null `context` as it is a **unit test** (not related to UI) and I therefore cannot create the `context` from such a test that does not call any @composable function.

### The decision I took
In order to keep the Toast functionality while being able to test the function, I choose to accept the passing of a null-context and to simply not display them when it is null (i.e when we are running the unit test).

### Changes

If I were to add a context check at the start it would imply that any rating submission would be denied if the context is null. I do not know if there are normal instances in which it could. Therefore I was reticent about doing so.

I can propose to add a **parameter** to the function to make the function silent, which would replace `context` checks by `silent` boolean checks if you want ?

Reply:
I don't understand why you can't check the context before the if-else statement.
You check the context in every single if-else statement, if the context is null to begin with you don't even need to check all the if-else statement.

Something like this
```
if (context != null){
  if (user == null) {
          Toast.makeText(context, "User not found, could not rate park", Toast.LENGTH_SHORT).show()
    } else if (park == null) {
          Toast.makeText(context, "Park not found, could not rate park", Toast.LENGTH_SHORT).show()
    }
}
```

> The reason for the null checking is that each Toast() instance require a context

or do you mean that each Toast have a different context and if it is then yes you would need to check context seperatly

Reply:
I sawed that in the else statement you had parkViewModel.addRating(park.pid, user.uid, starRating.toFloat()) that neede to be called at all time, so i understand why ou can't put context check at the start

Reply:
Each Toast need the same context. What I meant by not being able to check before the if-else statement is that it would modify the functionality of the handling of rating of parks (no longer only about Toasts). 

The only other possibility I see is to make the function return Int error codes which would be handled in the UI to display the Toasts, there would be no passing of context anymore but the complexity would increase a lot.